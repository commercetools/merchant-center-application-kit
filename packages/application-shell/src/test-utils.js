// This file allows consumers of the application-shell to test
// components rendered by the application shell.
// They should
// import { render } from ""@commercetools-frontend/application-shell/test-utils"
// and then use it together with react-testing-library.
import React from 'react';
import { Router } from 'react-router-dom';
import { render as rtlRender } from 'react-testing-library';
import { createMemoryHistory } from 'history';
import { IntlProvider } from 'react-intl';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import { MockedProvider as ApolloMockProvider } from 'react-apollo/test-utils';
import memoryAdapter from '@flopflip/memory-adapter';
import { Provider as StoreProvider } from 'react-redux';
import { ApplicationContextProvider } from '@commercetools-frontend/application-shell-connectors';
import { createReduxStore } from './configure-store';

// Reset memoryAdapter after each test, so that the next test accepts the
// defaultFlags param.
// This could also be moved into setup-test-framework, not sure which
// location is better for it.
afterEach(memoryAdapter.reset);

// These default values get merged with the values provided by the test from
// the call to "render"
const defaultUser = {
  id: 'user-id-1',
  email: 'sheldon.cooper@caltech.edu',
  firstName: 'Sheldon',
  lastName: 'Cooper',
  language: 'en',
  timeZone: 'Etc/UTC',
};

const defaultEnvironment = {
  frontendHost: 'localhost:3001',
  mcApiUrl: 'https://mc-api.commercetools.com',
  location: 'eu',
  env: 'production',
  cdnUrl: 'http://localhost:3001',
  servedByProxy: false,
};

const defaultProject = {
  key: 'test-with-big-data',
  version: 43,
  name: 'Test with big data',
  countries: ['de', 'en'],
  currencies: ['EUR', 'GBP'],
  languages: ['de', 'en-GB'],
};

const defaultPermissions = { canManageProject: true };

// Allow consumers of `render` to extend the defaults by passing an object
// or to completely omit the value by passing `null`
const mergeOptional = (defaultValue, value) =>
  value === null ? undefined : { ...defaultValue, ...value };

// This function renders any component within the application context, as if it
// was rendered inside <ApplicationShell />.
// The context is not completely set up yet, some things are missing:
//   - Tracking on context
//   - react-intl's information from addLocaleData
//   - possibly more that I'm not aware of right now
//
//  We can add these things as we go and when we need them.

// Inspired by
// https://github.com/kentcdodds/react-testing-library-course/blob/2a5b1560656790bb1d9c055fba3845780b2c2c97/src/__tests__/react-router-03.js
const render = (
  ui,
  {
    // react-intl
    locale = 'en',
    // Apollo
    mocks = [],
    addTypename = false,
    // react-router
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    // flopflip
    adpater = memoryAdapter,
    flags = {},
    // application-context
    environment,
    user,
    project,
    permissions = defaultPermissions,
    dataLocale = 'en',
    // forwarding to react-testing-library
    ...renderOptions
  } = {}
) => {
  const mergedUser = mergeOptional(defaultUser, user);
  const mergedProject = mergeOptional(defaultProject, project);
  const mergedEnvironment = mergeOptional(defaultEnvironment, environment);
  return {
    ...rtlRender(
      <IntlProvider locale={locale}>
        <ApolloMockProvider mocks={mocks} addTypename={addTypename}>
          <ConfigureFlopFlip adapter={adpater} defaultFlags={flags}>
            <ApplicationContextProvider
              user={mergedUser}
              project={mergedProject && { ...mergedProject, permissions }}
              environment={mergedEnvironment}
              projectDataLocale={dataLocale}
            >
              <Router history={history}>{ui}</Router>
            </ApplicationContextProvider>
          </ConfigureFlopFlip>
        </ApolloMockProvider>
      </IntlProvider>,
      renderOptions
    ),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
    // Adding user, project & environment so tests know about the merge results
    // Note that these objects do not resemble the application context, they are
    // only intended to communicate the test setup back to the tests.
    user: mergedUser,
    project: mergedProject,
    environment: mergedEnvironment,
  };
};

// Test setup for rendering with Redux
// We expose a sophisticated function because we plan to get rid of Redux
// Use this function only when your test actually needs Redux
const renderWithRedux = (
  ui,
  {
    // Consuemrs of renderWithRedux can use
    //   { store: createReduxStore({ requestsInFlight: null, .. }) }
    // to pass an initial state to Redux.
    store = createReduxStore(),
    ...renderOptions
  } = {}
) => ({
  ...render(<StoreProvider store={store}>{ui}</StoreProvider>, renderOptions),
  // adding `store` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  store,
});

// re-export everything
export * from 'react-testing-library';

export {
  // override render method of react-testing-library
  render,
  renderWithRedux,
  // the original "render" method of react-testing-library
  rtlRender,
};
