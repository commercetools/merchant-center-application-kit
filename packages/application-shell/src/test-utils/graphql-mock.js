import { graphql, buildClientSchema } from 'graphql';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { formatApolloErrors } from 'apollo-server-errors';
import mcSchemaIntrospectionResult from '../../../../schemas/mc.json';
import proxySchemaIntrospectionResult from '../../../../schemas/proxy.json';
import * as mcFixtures from './fixtures/mc';
import * as proxyFixtures from './fixtures/proxy';

const getSchemaIntrospectionFor = targetName => {
  switch (targetName) {
    case 'mc':
      return mcSchemaIntrospectionResult.data;
    case 'proxy':
      return proxySchemaIntrospectionResult.data;
    default:
      throw new Error(`Unknown target ${targetName}`);
  }
};

const schemas = {
  mc: buildClientSchema(getSchemaIntrospectionFor('mc')),
  proxy: buildClientSchema(getSchemaIntrospectionFor('proxy')),
};

const getDefaultFixturesFor = targetName => {
  switch (targetName) {
    case 'mc':
      return mcFixtures.defaultFixtures;
    case 'proxy':
      return proxyFixtures.defaultFixtures;
    default:
      throw new Error(`Unknown target ${targetName}`);
  }
};

const getMockedSchemaFor = (
  targetName,
  customFixtures = {},
  customResolvers = {}
) => {
  switch (targetName) {
    case 'mc': {
      const schema = schemas.mc;
      addMockFunctionsToSchema({
        schema,
        mocks: mcFixtures.createResolvers(
          customFixtures.mc,
          customResolvers.mc
        ),
        preserveResolvers: true,
      });
      return schema;
    }
    case 'proxy': {
      const schema = schemas.proxy;
      addMockFunctionsToSchema({
        schema,
        mocks: proxyFixtures.createResolvers(
          customFixtures.proxy,
          customResolvers.proxy
        ),
        preserveResolvers: true,
      });
      return schema;
    }
    default:
      throw new Error(`Unknown target ${targetName}`);
  }
};

const createGraphqlMockServer = (
  xhrMock,
  { fixtures, resolvers, onRequest } = {}
) => {
  const mockedSchemas = {
    mc: getMockedSchemaFor('mc', fixtures, resolvers),
    proxy: getMockedSchemaFor('proxy', fixtures, resolvers),
  };

  const processGraphQLRequest = async (targetName, req, res) => {
    const { query, variables } = JSON.parse(req.body());
    let response;
    try {
      response = await graphql(
        mockedSchemas[targetName],
        query,
        null,
        null,
        variables
      );
    } catch (error) {
      // https://github.com/apollographql/apollo-server/blob/61e69412c0aa88b358a2eef4ba635e8618db9946/packages/apollo-server-core/src/runHttpQuery.ts#L295-L297
      response = { error: formatApolloErrors([error]) };
    }
    // console.log(
    //   'response',
    //   JSON.stringify(response),
    //   targetName,
    //   query,
    //   variables
    // );
    return res.status(200).body(JSON.stringify(response));
  };
  const handleProxyGraphQLRequest = async (req, res) => {
    const { variables } = JSON.parse(req.body());
    const targetName = variables.target;
    return await processGraphQLRequest(targetName, req, res);
  };

  xhrMock.post('http://localhost:8080/graphql', async (req, res) => {
    if (onRequest) {
      const next = async () => {
        return await handleProxyGraphQLRequest(req, res);
      };
      return await onRequest(req, res, next);
    }
    return await handleProxyGraphQLRequest(req, res);
  });
  xhrMock.post(/\/api\/graphql$/, async (req, res) => {
    return await processGraphQLRequest('proxy', req, res);
  });
};

export { getDefaultFixturesFor, createGraphqlMockServer };
