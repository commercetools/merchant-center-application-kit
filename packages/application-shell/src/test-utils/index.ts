export * from './test-utils';
export * from './custom-views-utils';
