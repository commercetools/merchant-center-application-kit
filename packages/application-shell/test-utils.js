// this file is here because we want jest to use our non-compiled code to run tests
// if this file is missing, and you have a `module` or `main` that points to a non-existing file
// (ie, a bundle that hasn't been built yet) then jest will fail if the bundle is not yet built.
export * from './src/test-utils';
