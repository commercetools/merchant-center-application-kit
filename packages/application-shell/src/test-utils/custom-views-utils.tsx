import type { ReactNode } from 'react';
import { ApolloClient, type NormalizedCacheObject } from '@apollo/client';
import type { RenderResult } from '@testing-library/react';
import {
  CustomViewContextProvider,
  type TProviderProps,
} from '@commercetools-frontend/application-shell-connectors';
import {
  CUSTOM_VIEW_HOST_ENTRY_POINT_URI_PATH,
  type CustomViewData,
} from '@commercetools-frontend/constants';
import { TCustomViewType } from '../types/generated/settings';
import { renderAppWithRedux } from './test-utils';

const testCustomViewData: CustomViewData = {
  id: 'custom-view-id',
  defaultLabel: 'My View',
  labelAllLocales: [],
  url: 'https://my-view.com',
  type: TCustomViewType.CustomPanel,
  locators: ['products.product_details.general'],
  permissions: [
    {
      name: 'view',
      oAuthScopes: ['view_products'],
    },
  ],
};

type TRenderCustomViewParams = {
  locale: string;
  projectKey?: string;
  projectAllAppliedPermissions?: { name: string; value: boolean }[];
  customViewHostUrl?: string;
  customViewConfig?: Partial<CustomViewData>;
  apolloClient?: ApolloClient<NormalizedCacheObject>;
  environment?: Partial<TProviderProps<{}>['environment']>;
  user?: Partial<TProviderProps<{}>['user']>;
  children: ReactNode;
};

export const renderCustomView = (
  props: TRenderCustomViewParams
): RenderResult => {
  return renderAppWithRedux(
    <CustomViewContextProvider
      hostUrl={
        props.customViewHostUrl ??
        'https://mc.ct.com/my-project/products/product-id-1'
      }
      customViewConfig={{ ...testCustomViewData, ...props.customViewConfig }}
    >
      {props.children}
    </CustomViewContextProvider>,
    {
      apolloClient: props.apolloClient,
      locale: props.locale,
      project: {
        key: props.projectKey,
        allAppliedPermissions: props.projectAllAppliedPermissions || [],
      },
      environment: {
        ...(props.environment || {}),
        entryPointUriPath: CUSTOM_VIEW_HOST_ENTRY_POINT_URI_PATH,
      },
      user: props.user,
    }
  );
};
