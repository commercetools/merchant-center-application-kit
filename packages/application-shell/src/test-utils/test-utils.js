// This file allows consumers of the application-shell to test
// components rendered by the application shell.
// They should
// import { render } from ""@commercetools-frontend/application-shell/test-utils"
// and then use it together with react-testing-library.
import React from 'react';
import PropTypes from 'prop-types';
import { Router } from 'react-router-dom';
import invariant from 'tiny-invariant';
import { ApolloProvider } from 'react-apollo';
import { render as rtlRender } from 'react-testing-library';
import { createMemoryHistory } from 'history';
import { IntlProvider } from 'react-intl';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import { MockedProvider as ApolloMockProvider } from 'react-apollo/test-utils';
import memoryAdapter from '@flopflip/memory-adapter';
import { Provider as StoreProvider } from 'react-redux';
import { ApplicationContextProvider } from '@commercetools-frontend/application-shell-connectors';
import { NotificationsList } from '@commercetools-frontend/react-notifications';
import { DOMAINS } from '@commercetools-frontend/constants';
import { createTestMiddleware as createSdkTestMiddleware } from '@commercetools-frontend/sdk/test-utils';
import { createReduxStore } from '../configure-store';
import { createApolloClient } from '../configure-apollo';

// Reset memoryAdapter after each test, so that the next test accepts the
// defaultFlags param.
// This could also be moved into setup-test-framework, not sure which
// location is better for it.
afterEach(memoryAdapter.reset);

// These default values get merged with the values provided by the test from
// the call to "render"
const defaultUser = {
  id: 'user-id-1',
  email: 'sheldon.cooper@caltech.edu',
  firstName: 'Sheldon',
  lastName: 'Cooper',
  language: 'en',
  timeZone: 'Etc/UTC',
};

const defaultEnvironment = {
  frontendHost: 'localhost:3001',
  mcApiUrl: 'https://mc-api.commercetools.com',
  location: 'eu',
  env: 'production',
  cdnUrl: 'http://localhost:3001',
  servedByProxy: false,
};

const defaultProject = {
  key: 'test-with-big-data',
  version: 43,
  name: 'Test with big data',
  countries: ['de', 'en'],
  currencies: ['EUR', 'GBP'],
  languages: ['de', 'en-GB'],
  owner: {
    id: 'project-id-1',
  },
};

const defaultPermissions = { canManageProject: true };

// Allow consumers of `render` to extend the defaults by passing an object
// or to completely omit the value by passing `null`
const mergeOptional = (defaultValue, value) =>
  value === null ? undefined : { ...defaultValue, ...value };

const LoadingFallback = () => 'Loading...';
LoadingFallback.displayName = 'LoadingFallback';

const MockedApolloProvider = ({ children, mocks, addTypename }) => (
  <ApolloMockProvider mocks={mocks} addTypename={addTypename}>
    {children}
  </ApolloMockProvider>
);
MockedApolloProvider.displayName = 'MockedApolloProvider';
MockedApolloProvider.propTypes = {
  children: PropTypes.node.isRequired,
  mocks: PropTypes.array.isRequired,
  addTypename: PropTypes.bool.isRequired,
};

// This function renders any component within the application context, as if it
// was rendered inside <ApplicationShell />.
// The context is not completely set up yet, some things are missing:
//   - Tracking on context
//   - react-intl's information from addLocaleData
//   - possibly more that I'm not aware of right now
//
//  We can add these things as we go and when we need them.

// Inspired by
// https://github.com/kentcdodds/react-testing-library-course/blob/2a5b1560656790bb1d9c055fba3845780b2c2c97/src/__tests__/react-router-03.js
const render = (
  ui,
  {
    // react-intl
    locale = 'en',
    // Apollo
    mocks = [],
    addTypename = false,
    // react-router
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    // flopflip
    adpater = memoryAdapter,
    flags = {},
    // application-context
    environment,
    user,
    project,
    permissions = defaultPermissions,
    dataLocale = 'en',
    ApolloProviderComponent = MockedApolloProvider,
    // forwarding to react-testing-library
    ...renderOptions
  } = {}
) => {
  const mergedUser = mergeOptional(defaultUser, user);
  const mergedProject = mergeOptional(defaultProject, project);
  const mergedEnvironment = mergeOptional(defaultEnvironment, environment);
  return {
    ...rtlRender(
      <IntlProvider locale={locale}>
        <ApolloProviderComponent mocks={mocks} addTypename={addTypename}>
          <ConfigureFlopFlip adapter={adpater} defaultFlags={flags}>
            <ApplicationContextProvider
              user={mergedUser}
              project={mergedProject && { ...mergedProject, permissions }}
              environment={mergedEnvironment}
              projectDataLocale={dataLocale}
            >
              <Router history={history}>
                <React.Suspense fallback={<LoadingFallback />}>
                  {ui}
                </React.Suspense>
              </Router>
            </ApplicationContextProvider>
          </ConfigureFlopFlip>
        </ApolloProviderComponent>
      </IntlProvider>,
      renderOptions
    ),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
    // Adding user, project & environment so tests know about the merge results
    // Note that these objects do not resemble the application context, they are
    // only intended to communicate the test setup back to the tests.
    user: mergedUser,
    project: mergedProject,
    environment: mergedEnvironment,
  };
};

// Test setup for rendering with Redux
// We expose a sophisticated function because we plan to get rid of Redux
// Use this function only when your test actually needs Redux
const renderWithRedux = (
  ui,
  {
    // Consumers of renderWithRedux can use
    //   { store: createReduxStore({ requestsInFlight: null, .. }) }
    // to pass an initial state to Redux.
    store = undefined,
    // The store option is kept around to keep the API open as not all use-cases
    // are known yet. Meanwhile storeState and sdkMocks provide convenient ways
    // to work with the redux store.
    storeState = undefined,
    // renderWithRedux supports mocking requests made through the sdk
    // middleware. The approach is inspired by ApolloMockProvider.
    // You can pass responses for specific actions like
    //   renderWithRedux(ui, {
    //     sdkMocks: [
    //       {
    //         action: { type: 'SDK', payload: {}},
    //         response: { foo: true },
    //       }
    //     ]
    //   })
    // You can also fake a failing request using
    //   renderWithRedux(ui, {
    //     sdkMocks: [
    //       {
    //         action: { type: 'SDK', payload: {}},
    //         error: new Error('foo'),
    //       }
    //     ]
    //   })
    // Note that each response will only be used once. When multiple responses
    // are provided for identical actions, then they are used in the order
    // they are provided in.
    sdkMocks = [],
    ...renderOptions
  } = {}
) => {
  invariant(
    !(store && storeState),
    'test-utils: You provided store and storeState. Provide one of them only.'
  );
  invariant(
    !(store && sdkMocks.length > 0),
    'test-utils: You provided store and sdkMocks. Provide one of them only.'
  );

  // Determine the redux store to use in tests.
  // - When the user passed in a "store", we use that store and ignore
  //   sdkMocks and storeState.
  // - When the user passed in no sdkMocks, we create a store using the
  //   provided storeState. If storeState is undefined, then the defaults kick
  //   in anyways.
  // - Lastly, when sdkMocks were provided (and no store was provided), we
  //   create a store which applies a special middleware to allow mocking sdk
  //   responses. We further initialize the store with the provided storeState.
  //   If storeState is undefined, then the defaults kick in anyways.
  const reduxStore = (() => {
    if (store) return store;

    if (sdkMocks.length === 0) return createReduxStore(storeState);

    const testingMiddleware = createSdkTestMiddleware(sdkMocks);
    return createReduxStore(storeState, [testingMiddleware]);
  })();

  return {
    ...render(
      <StoreProvider store={reduxStore}>
        <div>
          <NotificationsList domain={DOMAINS.GLOBAL} />
          <NotificationsList domain={DOMAINS.PAGE} />
          <NotificationsList domain={DOMAINS.SIDE} />
          {ui}
        </div>
      </StoreProvider>,
      renderOptions
    ),
    // adding `store` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    store: reduxStore,
  };
};

// Renders UI without mocking ApolloProvider
const experimentalRender = (ui, renderOptions) => {
  const client = createApolloClient();
  const RealApolloProvider = ({ children }) => (
    <ApolloProvider client={client}>{children}</ApolloProvider>
  );
  RealApolloProvider.displayName = 'RealApolloProvider';
  RealApolloProvider.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return renderWithRedux(ui, {
    ...renderOptions,
    ApolloProviderComponent: RealApolloProvider,
  });
};

// re-export everything
export * from 'react-testing-library';

export {
  // override render method of react-testing-library
  render,
  renderWithRedux,
  experimentalRender,
  // the original "render" method of react-testing-library
  rtlRender,
};
