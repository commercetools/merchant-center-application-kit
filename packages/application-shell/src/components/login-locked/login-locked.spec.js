import React from 'react';
import { renderApp, waitForElement } from '../../test-utils';
import LockedAccount from './async';

it('renders maintenance page for locked account', async () => {
  const { getByText, queryByText } = renderApp(<LockedAccount />);
  await waitForElement(() => getByText(/Account Locked/i));
  expect(queryByText(/reset your password/i)).toBeDefined();
  expect(queryByText(/Help Desk/i)).toBeDefined();
});
