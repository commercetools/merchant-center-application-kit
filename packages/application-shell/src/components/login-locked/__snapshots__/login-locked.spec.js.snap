// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpDeskLink rendering should match snapshot 1`] = `
<a
  href="https://support.commercetools.com"
  target="_blank"
>
  <FormattedMessage
    defaultMessage="Help Desk"
    id="LoginLocked.helpDeskLink"
    values={Object {}}
  />
</a>
`;

exports[`LockedAccount rendering rendering should match snapshot 1`] = `
<ServicePageResponseLayout
  imageSrc="test-file-stub"
  paragraph1={
    <FormattedMessage
      defaultMessage="For security reasons, your account has been locked because there were too many invalid log in attempts. You can {resetPasswordLink} or contact the commercetools {helpDeskLink}"
      id="LoginLocked.paragraph1"
      values={
        Object {
          "helpDeskLink": <HelpDeskLink />,
          "resetPasswordLink": <ResetPasswordLink />,
        }
      }
    />
  }
  paragraph2={
    <TextDetail>
      2018 Â© commercetools
    </TextDetail>
  }
  title={
    <FormattedMessage
      defaultMessage="Account Locked"
      id="LoginLocked.title"
      values={Object {}}
    />
  }
/>
`;

exports[`ResetPasswordLink rendering rendering should match snapshot 1`] = `
<GetApplicationState
  render={[Function]}
/>
`;
